# [LTFS Data Science FinHack ( ML Hackathon)](https://datahack.analyticsvidhya.com/contest/ltfs-datascience-finhack-an-online-hackathon/)

### Notes -
I have used [modin](https://github.com/modin-project/modin) package which speeds up pandas library.

#### To install modin
`!pip install modin` then also do a `!pip install setproctitle`<br>
drop the initial '!' from above commands if you run in a Terminal. I am using [Azure Notebook](https://notebooks.azure.com)

In [17]:
import numpy as np
import pandas as pd

In [18]:
pd.set_option('display.max_columns', None)

In [19]:
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

### Reading the data from disk

In [20]:
df_train = pd.read_csv('train.csv')
df_test = pd.read_csv('test.csv')
sample_submission = pd.read_csv('sample_submission.csv')

#### Lets check the data quality first

In [21]:
df_train.shape

(233154, 41)

In [22]:
# Percentage of null values in each feature
def get_nulls(X):
    key_string = 'Null '+'/ '+str(X.shape[0])
    temp = pd.DataFrame({key_string: X.isnull().sum(),
                         'Null %age': round(X.isnull().sum()/ X.shape[0]*100,2)})
    temp = temp.loc[temp[key_string] > 0,:]
    return temp

In [23]:
get_nulls(df_train)

Null %age  Null / 233154
Employment.Type       3.29           7661

In [24]:
get_nulls(df_test)

Null %age  Null / 112392
Employment.Type       3.06           3443

Among 41 columns only 1 columns **Employment.type** is having null values. Also, only about 3% of rows have nulls. But if we see it more carefully, since our dataset is pretty large the **3% is actually 7661 rows**.<br> Entire model can be built using such data. So why drop these rows. We can use **segmentation** here easily.

#### Lets do some EDA

In [25]:
df_train.head()

UniqueID  disbursed_amount  asset_cost    ltv  branch_id  supplier_id  \
0    420825             50578       58400  89.55         67        22807   
1    537409             47145       65550  73.23         67        22807   
2    417566             53278       61360  89.63         67        22807   
3    624493             57513       66113  88.48         67        22807   
4    539055             52378       60300  88.39         67        22807   

   manufacturer_id  Current_pincode_ID Date.of.Birth Employment.Type  \
0               45                1441      01-01-84        Salaried   
1               45                1502      31-07-85   Self employed   
2               45                1497      24-08-85   Self employed   
3               45                1501      30-12-93   Self employed   
4               45                1495      09-12-77   Self employed   

  DisbursalDate  State_ID  Employee_code_ID  MobileNo_Avl_Flag  Aadhar_flag  \
0      03-08-18         6              1998                  1            1   
1      26-09-18         6              1998                  1            1   
2      01-08-18         6              1998                  1            1   
3      26-10-18         6              1998                  1            1   
4      26-09-18         6              1998                  1            1   

   PAN_flag  VoterID_flag  Driving_flag  Passport_flag  PERFORM_CNS.SCORE  \
0         0             0             0              0                  0   
1         0             0             0              0                598   
2         0             0             0              0                  0   
3         0             0             0              0                305   
4         0             0             0              0                  0   

  PERFORM_CNS.SCORE.DESCRIPTION  PRI.NO.OF.ACCTS  PRI.ACTIVE.ACCTS  \
0   No Bureau History Available                0                 0   
1                 I-Medium Risk                1                 1   
2   No Bureau History Available                0                 0   
3              L-Very High Risk                3                 0   
4   No Bureau History Available                0                 0   

   PRI.OVERDUE.ACCTS  PRI.CURRENT.BALANCE  PRI.SANCTIONED.AMOUNT  \
0                  0                    0                      0   
1                  1                27600                  50200   
2                  0                    0                      0   
3                  0                    0                      0   
4                  0                    0                      0   

   PRI.DISBURSED.AMOUNT  SEC.NO.OF.ACCTS  SEC.ACTIVE.ACCTS  SEC.OVERDUE.ACCTS  \
0                     0                0                 0                  0   
1                 50200                0                 0                  0   
2                     0                0                 0                  0   
3                     0                0                 0                  0   
4                     0                0                 0                  0   

   SEC.CURRENT.BALANCE  SEC.SANCTIONED.AMOUNT  SEC.DISBURSED.AMOUNT  \
0                    0                      0                     0   
1                    0                      0                     0   
2                    0                      0                     0   
3                    0                      0                     0   
4                    0                      0                     0   

   PRIMARY.INSTAL.AMT  SEC.INSTAL.AMT  NEW.ACCTS.IN.LAST.SIX.MONTHS  \
0                   0               0                             0   
1                1991               0                             0   
2                   0               0                             0   
3                  31               0                             0   
4                   0               0                             0   

   D

In [26]:
df_train.shape

(233154, 41)

In [27]:
df_test.head()

UniqueID  disbursed_amount  asset_cost    ltv  branch_id  supplier_id  \
0    655269             53478       63558  86.54         67        22807   
1    723482             55513       63163  89.45         67        22807   
2    758529             65282       84320  79.93         78        23135   
3    763449             46905       63896  76.58         78        17014   
4    708663             51428       63896  86.08         78        17014   

   manufacturer_id  Current_pincode_ID Date.of.Birth Employment.Type  \
0               45                1497      01-01-74        Salaried   
1               45                1497      20-05-85   Self employed   
2               86                2071      14-10-95        Salaried   
3               45                2070      01-06-73   Self employed   
4               45                2069      01-06-72        Salaried   

  DisbursalDate  State_ID  Employee_code_ID  MobileNo_Avl_Flag  Aadhar_flag  \
0      03-11-18         6              1998                  1            1   
1      20-11-18         6              1998                  1            1   
2      29-11-18         4              1646                  1            1   
3      29-11-18         4              1646                  1            1   
4      17-11-18         4              1646                  1            1   

   PAN_flag  VoterID_flag  Driving_flag  Passport_flag  PERFORM_CNS.SCORE  \
0         0             0             0              0                  0   
1         0             0             0              0                749   
2         1             0             0              0                  0   
3         0             0             0              0                 14   
4         0             0             0              0                  0   

  PERFORM_CNS.SCORE.DESCRIPTION  PRI.NO.OF.ACCTS  PRI.ACTIVE.ACCTS  \
0   No Bureau History Available                0                 0   
1               C-Very Low Risk                2                 1   
2   No Bureau History Available                0                 0   
3  Not Scored: Only a Guarantor                1                 1   
4   No Bureau History Available                0                 0   

   PRI.OVERDUE.ACCTS  PRI.CURRENT.BALANCE  PRI.SANCTIONED.AMOUNT  \
0                  0                    0                      0   
1                  0                43898                  48780   
2                  0                    0                      0   
3                  1               132480                 255000   
4                  0                    0                      0   

   PRI.DISBURSED.AMOUNT  SEC.NO.OF.ACCTS  SEC.ACTIVE.ACCTS  SEC.OVERDUE.ACCTS  \
0                     0                0                 0                  0   
1                 48780                0                 0                  0   
2                     0                0                 0                  0   
3                255000                0                 0                  0   
4                     0                0                 0                  0   

   SEC.CURRENT.BALANCE  SEC.SANCTIONED.AMOUNT  SEC.DISBURSED.AMOUNT  \
0                    0                      0                     0   
1                    0                      0                     0   
2                    0                      0                     0   
3                    0                      0                     0   
4                    0                      0                     0   

   PRIMARY.INSTAL.AMT  SEC.INSTAL.AMT  NEW.ACCTS.IN.LAST.SIX.MONTHS  \
0                   0               0                             0   
1                5605               0                             1   
2                   0               0                             0   
3                   0               0                             0   
4                   0               0                             0   

   D

In [28]:
df_test.shape

(112392, 40)

#### Model Building

In [13]:
y_train = df_train.pop('loan_default')
X_train = df_train
X_test  = df_test

In [15]:
from sklearn.model_selection import train_test_split

In [16]:
X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size=0.33, random_state=42, stratify=y_train)